### DISCUSSION TITLE
How to define a udf in python with bytes (blob) as input and map or struct as output?

### DISCUSSION THREAD

**Discussion by suiluj** [duckdb/duckdb#15343] [Q&A] [answered]
Hello,

I really like the udf feature in duckdb.
Now I would like to try some *more advanced* UDF but I am not sure how to define the udf. Perhaps this is just not supported

### Goal

- define a udf with bytes as input and a structured schema output (multiple fields as part of a map or struct)

#### example

```python
def pdf_process_settlement(pdf_bytes):
    s = settlement.Settlement.from_pdf_bytes(pdf_bytes=pdf_bytes)
    s_dict = s.to_dict()
    # s_dict contains some information inside a dictionary
    # the keys contain static types (i use this dictionary for pyarrow table creation from pylist often times
    return_dict = {
        "bank_account": s_dict["bank_account"],
        "currency": s_dict["currency"]
    }
    return return_dict

duckdb.create_function("pdf_process_settlement", pdf_process_settlement, [BLOB], {'bank_account': VARCHAR, 'currency': VARCHAR})

query = """
select
    *
    ,pdf_process_settlement(pdf_bytes) as processed_pdf
from document_billings_settlements
where settlement_type = 'buy'
limit 5
"""
df = con.sql(query=query).pl()
```

I got this error:

```
TypeError: create_function(): incompatible function arguments. The following argument types are supported:
    1. (name: str, function: Callable, parameters: object = None, return_type: duckdb.duckdb.typing.DuckDBPyType = None, *, type: duckdb.duckdb.functional.PythonUDFType = <PythonUDFType.NATIVE: 0>, null_handling: duckdb.duckdb.functional.FunctionNullHandling = <FunctionNullHandling.DEFAULT: 0>, exception_handling: duckdb.duckdb.PythonExceptionHandling = <PythonExceptionHandling.DEFAULT: 0>, side_effects: bool = False, connection: duckdb.DuckDBPyConnection = None) -> duckdb.DuckDBPyConnection
```

My guess is that this type of udf output is just not supported right now.  
But I would really prefer to define my udf with a schema that uses native types.  
Of course an alternative would be to return a json varchar and then map the json to a schema in sql but I do not want to convert my key values between string and back to its correct type.

**Comment by Tishj** [comment, marked as answer]
```py
import duckdb
from duckdb.typing import (
	DuckDBPyType,
	VARCHAR,
	BLOB
)

con = duckdb.connect()

def pdf_process_settlement(pdf_bytes):
    s = pdf_bytes
    return_dict = {
        "bank_account": 'hello',
        "currency": 'world'
    }
    return return_dict

return_type = DuckDBPyType({'bank_account': str, 'currency': str})
con.create_function(name="pdf_process_settlement", function=pdf_process_settlement, parameters=[BLOB], return_type=return_type)

query = """
select
	pdf_process_settlement(encode('\xaa\xff\xaa')) as processed_pdf
"""
con.sql(query).show()

```

**Reply by Tishj** [reply]
I didn't verify, but I think the use of `{'bank_account': VARCHAR, 'currency': VARCHAR}` is the problem in your example
(besides registering the function on `duckdb` and then using `con`, which does not have the registered function)

As the error says, it wants `return_type` to be a DuckDBPyType, it will attempt an implicit conversion, but this failed
<https://github.com/duckdb/duckdb/pull/15346> should fix this issue I believe

**Reply by suiluj** [reply]
ah oops my bad with the wrong `duckdb.create_function` and then `con.sql`. I changed my example code for this issue mixed up a few things. :D 

Thanks for the hint with `DuckDBPyType`

```python
from duckdb.typing import DuckDBPyType
DuckDBPyType({'bank_account': str, 'currency': str})
```
I was not aware of this syntax.

Will give it a try! 👍

**Reply by suiluj** [reply]
@Tishj ah thanks by the way. it works perfectly! :)
